docs/error-handling.md

```diff
@@ -0,0 +1,190 @@
+# Error handling and logging
+
+## Logging errors
+
+The principle is that errors are logged when they are handled. If you
+just propagate an error to the caller in a function, you don't need to
+log it; the caller will. But if you "swallow" an error in a function,
+you *must* log it (if it needs to be logged at all).
```

Comment:
> As I mentioned above, this has the tendency to divorce errors from the context of the spans they were created in.

Huh, you're right! For some reason, my mental model has been that an anyhow::Error captures the span that it was created in, but it doesn't.

There is https://docs.rs/tracing-error/latest/tracing_error/ for capturing the current span, but anyhow won't use it automatically.

Reply:
Raced to comment on `SpanTrace` in topmost thread: https://github.com/neondatabase/neon/pull/2681#issuecomment-1288554825 but this thread seems better suited.

After some [experimentation](https://gist.github.com/koivunej/6f6ac896ddea7f6be01245bb2ca56097) (includes example output in comment), `color-eyre` is required to format and `tracing-error` are required to capture the `SpanTrace`. The SPANTRACE header comes from [`color-spantrace`](https://github.com/yaahc/color-spantrace/blob/1a52aac2b55a05efc64807f06b91414550805af8/src/lib.rs#L353).

Reply:
That's very cool! Seeing that, I wonder if it's possible to capture the `Span` object itself, so that logging errors can temporarily re-enter all of the necessary spans instead. If not, I think `SpanTrace` is still a great solution to this :)